### PR TITLE
Fix scheduler thread wakeups

### PR DESCRIPTION
**Pull request by veerajsk**
It was possible for the scheduler thread to wait on its condition
variable while the condition variable was notified during both
termination & addition of a job to the job queue due to race conditions
resulting from the unlocked manipulation of the running flag & the job
queue. Fix this by grabbing the wait mutex before modifying said
variables.

### CODE REVIEW

File: alica_engine/include/engine/scheduler/Scheduler.h [hunk 1]
```diff
@@ -66,9 +69,12 @@ class Scheduler
 
     void terminate()
     {
-        _running = false;
-
+        {
+            std::unique_lock<std::mutex> lock(_mutex);
+            _running = false;
+        }
         _cv.notify_one();
```
Comment (xffox) [inline alica_engine/include/engine/scheduler/Scheduler.h:76]: Hm, i think you might still loose the `_cv` notification on around line 100 before the lock. Maybe then you need to check the values before the `wait()`, but under the lock.

Comment (veerajsk) [reply alica_engine/include/engine/scheduler/Scheduler.h:76]: that's okay, since `running` will be false in that case & we will never wait on the condition variable

Comment (veerajsk) [reply alica_engine/include/engine/scheduler/Scheduler.h:76]: the predicate passed to the wait() function is called before going into wait but under the lock

Comment (xffox) [reply alica_engine/include/engine/scheduler/Scheduler.h:76]: right

**Review by xffox — APPROVED** [review]

**Comment by xffox** [comment]
The job queue is not under the lock, are there performance concerns?

**Comment by veerajsk** [comment]
> The job queue is not under the lock, are there performance concerns?

The queue itself is multi thread safe i.e. it has its own internal lock that it uses to guard access

**Review by abhi1287 — APPROVED** [review]

**Comment by xffox** [comment]
> > The job queue is not under the lock, are there performance concerns?
> 
> The queue itself is multi thread safe i.e. it has its own internal lock that it uses to guard access

Right, i mean with one more lock it's slower, would it affect someone?

File: alica_engine/include/engine/scheduler/Scheduler.h [hunk 2]
```diff
@@ -53,7 +53,10 @@ class Scheduler
     JobId schedule(JobCb jobCb, std::optional<AlicaTime> repeatInterval = std::nullopt)
     {
         JobId jobId = _nextJobId++;
-        _jobQueue.emplace(jobId, std::move(jobCb), std::move(repeatInterval));
+        {
+            std::unique_lock<std::mutex> lock(_mutex);
+            _jobQueue.emplace(jobId, std::move(jobCb), std::move(repeatInterval));
```
Comment (abhi1287) [inline alica_engine/include/engine/scheduler/Scheduler.h:58]: don't see why is this necessary? as a general thumb rule we should avoid acquiring mutex inside mutex.

Comment (veerajsk) [reply alica_engine/include/engine/scheduler/Scheduler.h:58]: so the sequence that would fail I think is:
- Scheduler thread sees that's the `_jobQueue` is empty inside the `wait()` predicate i.e. `pop()` returns an empty optional
- Another thread adds a job to `_jobQueue` & calls `notify_one()`
- Scheduler thread misses the notify because it is still executing the `wait()` predicate & ends up waiting on the condition variable

Regarding the double lock, I am currently checking if we can get rid of the internal mutex in the queue.

Comment (veerajsk) [reply alica_engine/include/engine/scheduler/Scheduler.h:58]: I got rid of the job queue's mutex

**Comment by veerajsk** [comment]
> > > The job queue is not under the lock, are there performance concerns?
> > 
> > 
> > The queue itself is multi thread safe i.e. it has its own internal lock that it uses to guard access
> 
> Right, i mean with one more lock it's slower, would it affect someone?

I looked a bit more into the code & got rid of the internal mutex in the job queue, since we need to acquire the condition variable's mutex in any case while adding to the queue. So I used it for protecting access to the queue everywhere.

**Review by abhi1287 — APPROVED** [review]

File: alica_engine/include/engine/scheduler/Scheduler.h [hunk 3]
```diff
@@ -89,8 +96,8 @@ class Scheduler
     void run()
     {
         while (_running) {
-            auto job = _jobQueue.pop();
-            if (!job) {
+            std::optional<Job> job;
```
Comment (xffox) [inline alica_engine/include/engine/scheduler/Scheduler.h:99]: Can you also add `optional` include for this?

**Review by xffox — APPROVED** [review]
